[unitaryHACK] Update interface introduction

ryanhill1 opened the pull request:
**Context:**

Outdated interface introduction

**Description of the Change:**

In the [interface introduction](https://pennylane.readthedocs.io/en/stable/introduction/interfaces/torch.html), updated the sections on converting an existing QNode so that the functions `to_torch()` and `to_tf()` alter the QNode in place instead of returning a new QNode. Also, updated each interface's introduction to show creating a QNode with an interface via the class constructor:

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

#1155

ryanhill1 commented on doc/introduction/interfaces/tf.rst:
```diff
@@ -5,10 +5,10 @@ TensorFlow interface
-executed by the :func:`~.interfaces.tf.to_tf` function that returns the
-new quantum node object.
+either by using the ``interface='tf'`` flag in the QNode decorator or QNode class constructor, or by
+calling the :meth:`QNode.to_tf() <pennylane.qnodes.JacobianQNode.to_tf>` method. Internally, the
```
Is the reference <pennylane.qnodes.JacobianQNode.to_tf> correct? Or should it be <pennylane.qnode.QNode.to_torch> ?

albi3ro replied:
The second one.  No such thing as a `JacobianQNode` (I think?).

josh146 replied:
Yep, `JacobianQNode` is old and has long been removed 🙂

ryanhill1 commented on doc/introduction/interfaces/torch.rst:
```diff
@@ -68,15 +62,31 @@ use different classical interfaces:
 
->>> qnode1_torch = qnode1.to_torch()
->>> qnode1_torch
-<QNode: device='default.qubit', func=circuit, wires=2, interface=PyTorch>
```
Running the cells as they were before doesn't seem to give the `<QNode: device='default.qubit', func=circuit, wires=2, interface=PyTorch>` output. Instead, I get something like `<pennylane.qnode.QNode at 0x7ff01284d370>`.  Do you know the root of this inconsistency?

josh146 replied:
Good catch, we should definitely add the `__repr__` method back to the QNode - it must have gotten lost in a recent refactor. @albi3ro, could you open an issue for this?

albi3ro commented on doc/introduction/interfaces/tf.rst:
```diff
@@ -5,10 +5,10 @@ TensorFlow interface
-new quantum node object.
+either by using the ``interface='tf'`` flag in the QNode decorator or QNode class constructor, or by
+calling the :meth:`QNode.to_tf() <pennylane.qnodes.JacobianQNode.to_tf>` method. Internally, the
+translation is executed by the :func:`~.interfaces.torch.to_torch` function that dynamically modifies
```
I don't think we need to mention internal details here.

As a note, the interfaces are mix-in classes that can get added onto a tape by `~.interfaces.tf.TFInterface.apply`, but even most developers are not going to be using that information.

albi3ro reviewed (changes requested):
Thanks @ryanhill1 !

I left a lot of comments, but I can get pretty opinionated about writing. The comments about flow and phrasing are more optional potential improvements than required changes.

The things that do need to get changed are:
* correct function links and names
* section titles rendering well in build

Potential improvements:
* Remove discussion of implementation details
* explanation of when we may need to convert existing qnode
* move some text around and rephrase some other stuff

ryanhill1 commented:
Thank you so much for your great feedback! I'll get right on those changes

albi3ro commented:
Looks much better 👍

josh146 commented:
@ryanhill1 is this draft PR now final? 🙂

ryanhill1 commented:
@josh146 Just about! Should I add similar updates to the [JAX interface page](https://github.com/PennyLaneAI/pennylane/blob/master/doc/introduction/interfaces/jax.rst)? I was holding off on that one until I got the PyTorch and TensorFlow ones right so I could just work off of those templates.

albi3ro commented on doc/introduction/interfaces/tf.rst:
```diff
@@ -24,61 +23,78 @@ Tensorflow is imported as follows:
+<QNode: device='default.mixed', func=circuit1, wires=2, interface=TensorFlow>
+
+``qnode1()`` is now a TensorFlow-capable QNode, as well. Internally, the :meth:`~.QNode.to_tf`
+method uses the :func:`~.TFQNode` function to do the conversion.
```
Internally, the method uses [`~.interfaces.tf.TFInterface.apply`](https://github.com/PennyLaneAI/pennylane/blob/0965bb0c97c4de2dc2408751ca3a61dd47074264/pennylane/interfaces/tf.py#L225) to do the conversion.

```suggestion
method uses the :meth:`~.interfaces.tf.TFInterface.apply` function to do the conversion.
```

Check that actually works.

albi3ro commented on doc/introduction/interfaces/torch.rst:
```diff
@@ -21,66 +20,84 @@ new quantum node object.
+
+``qnode1()`` is now a PyTorch-capable QNode, as well. Internally, the
+:meth:`QNode.to_torch <qnode.QNode.to_torch>` method uses the
 :func:`TorchQNode <interfaces.torch.TorchQNode>` function to do the conversion.
```
```suggestion
:meth:`TorchInterface.apply <~.interfaces.torch.TorchInterface.apply>` function to do the conversion.
```

Looks like this function needs to get updated here too.

albi3ro reviewed (changes requested):
Looks great!

I left some comments about fixing the method names for the internal implementation, but after that it should be good to go :)

Above and beyond what I was initially wanting; a definite improvement.

albi3ro commented on doc/introduction/interfaces/tf.rst:
```diff
@@ -24,61 +23,78 @@ Tensorflow is imported as follows:
+<QNode: device='default.mixed', func=circuit1, wires=2, interface=TensorFlow>
+
+``qnode1()`` is now a TensorFlow-capable QNode, as well. Internally, the :meth:`~.QNode.to_tf`
+method uses the :meth:`~.interfaces.tf.TFInterface.apply` function to do the conversion.
```
```suggestion
method uses the :meth:`TFInterface.apply <~.interfaces.tf.TFInterface.apply>` function to do the conversion.
```

This was rendering as simply `apply`, so maybe we should try this instead. Neither was linking up to source code, but I'm not good enough at sphinx to figure out why.  Seems like maybe the interfaces are hidden internal objects that don't show up in the documentation.

I still don't think users need this information, especially if the method doesn't even show up in the documentation, but the text should at least be something more informative  than `apply`.

josh146 replied:
Perhaps it could be moved to a call out/hidden dropdown for advanced users?

albi3ro reviewed (changes requested):
Last tiny detail. 

Just make sure to check the build by going to the PR's checks, then `docs/readthedocs.org:pennylane` and clicking on `details`.

josh146 commented:
> @josh146 Just about! Should I add similar updates to the JAX interface page? I was holding off on that one until I got the PyTorch and TensorFlow ones right so I could just work off of those templates.

@ryanhill1 adding a similar update to the JAX interface page would be great (and this is something I know @albi3ro has also been thinking about!).

Best to keep it as a separate PR though. Usually it is best to have a separate PR per ‘task’/‘feature’/‘bug fix’, to help with the code review process.

ryanhill1 commented:
@josh146 @albi3ro Ok, I just deleted the lines about the internal use of the `apply` function in both the PyTorch and TensorFlow pages. I think you're right that it's not needed. I reviewed the builds in `docs/readthedocs.org:pennylane` and both files look good. And I also updated the CHANGELOG file. So everything should be ready for code review now!

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -344,14 +344,16 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 * Removes occurrences of the deprecated device argument ``analytic`` from the documentation.
   [(#1261)](https://github.com/PennyLaneAI/pennylane/pull/1261)
+  
+* Updated PyTorch and TensorFlow interface introductions [(#1333)](https://github.com/PennyLaneAI/pennylane/pull/1333)
```
```suggestion
* Updated PyTorch and TensorFlow interface introductions.
  [(#1333)](https://github.com/PennyLaneAI/pennylane/pull/1333)
```
(just to match the convention of the other entries)

josh146 reviewed (commented):
Awesome, thanks @ryanhill1! 💪 This looks great, and thanks for updating some of the (quite outdated) code snippets and links.

I did a quick skim through for rendering and ReST formatting, and all looks good on my end. I've tagged @albi3ro for a final review, and then should be ready for merging 🙂

albi3ro reviewed (approved):
Thanks! All good on my end :)
